### PR TITLE
Add support for the NVIDIA HPC SDK

### LINKED ISSUE #17930: PGI compilers integration in NVIDIA HPC SDK
NVIDIA released its HPC SDK today. It includes the ex-PGI compilers which have been rebranded `nvc`, `nvc++` and `nvfortran`.

How do you plan to accommodate this rebranding? Should we keep using the NVIDIA compilers as PGI compilers or should it be new compilers (which might break existing packages since "%pgi" wouldn't catch those although it will probably be needed in most cases)?

### PR DESCRIPTION
**Pull request by samcmill**
Add support for the [NVIDIA HPC SDK](https://developer.nvidia.com/hpc-sdk).  This adds a new nvhpc compiler class and a nvhpc package.

This PR is submitted as "official" support from NVIDIA for the HPC SDK in Spack.

A simple test to verify:
```
$ spack install nvhpc
$ eval $(spack load --sh nvhpc)
$ spack compiler find
==> Added 1 new compiler to /home/smcmillan/.spack/linux/compilers.yaml
    nvhpc@20.9
$ spack install netcdf-c%nvhpc ^openmpi%gcc ^m4%gcc
```

I have verified this is working correctly on Arm, Power, and X86.

A forthcoming PR will enable multiple packages to build correctly with the HPC SDK, removing the need to use `^openmpi%gcc ^m4%gcc`.  

The nvhpc compiler class and package are a clean break from the PGI compiler.  The NVIDIA HPC SDK is more than simply a rebranding of the PGI compiler.  Also I did not want to disrupt any existing builds that use PGI. 

The HPC SDK also includes CUDA.  A future enhancement would be to add a cuda variant.

Closes #17930
Closes #18261 
Closes #19139 
Closes #19162

cc @scheibelp @tgamblin

### CODE REVIEW

**Comment by scheibelp** [comment]
I have updated the PR description to mention that this closes #18261, #19139, and #19162. This PR appears to include the updates of those other 3, plus some additional changes.

@ajkotobi @Nischay-Pro I was going to be looking through all of these PRs later today but I was curious if this PR looked suitable for your usage.

File: lib/spack/spack/test/compilers/detection.py
Comment (scheibelp) [inline lib/spack/spack/test/compilers/detection.py]: Flake is complaining about this, it looks like it is "off-by-one" (needs to move to the left by 1 space).

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: You can add a `conflicts` directive at the `Package` level like:

```
class Nvhpc(Package):
  variant(...)
  conflicts('+network', when='+single', msg="You cannot choose both a network install and a system install")
```

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/nvhpc/package.py:83]: If you set `NVHPC_INSTALL_LOCSAL_DIR` to be `prefix`, then all the places that use `setup_run_environment` will already set up path variables like `PATH`, `CPATH`, etc. appropriately so you wouldn't have to set them there (although you would still want to set compiler exe variables like `CC` there).

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: (question) what is a network install?

**Review by scheibelp — COMMENTED** [review]
So far I have only looked at the `nvhpc` package file (I will be able to continue with looking at this PR tomorrow). I have a question and a couple suggestions.

Compared to https://github.com/spack/spack/pull/18261 (which also adds a Nvidia HPC SDK package) this has added support for installs on different architectures (addressing a shortcoming in Spack relating to packages which require different source downloads when building on different architectures).

#18261 also includes a mention of MPI support - I'm curious if that is enabled conditionally or if it is always available (not that it has to be exposed in this PR).

**Comment by adamjstewart** [comment]
The fact that 3 different contributors have opened up 4 different PRs in such a short period of time is a sign that people really want this package! That's great, but we also need to work together on this. I would like to see 1 PR instead of 4, that all contributors agree is closest to completion.

**Comment by Nischay-Pro** [comment]
My current use case requires me to use CUDA 10.2 since nsight does not support profiling on Pascal series. Would it be possible to add the CUDA variant flag? You can refer to my PR #19139 which has this feature already.

**Comment by ajkotobi** [comment]
It would be great to cover `cuda` variant and have `mpi` under environment setup section as conditional so to give more options during compilation as in [#18261](https://github.com/spack/spack/pull/18261), it was suggested to be there for completeness.

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py:83]: I may be misunderstanding your comment... there is an additional directory level in the HPC SDK.  `prefix/{bin,lib,...}` do not exist.  For example, the compilers exist in `prefix/compilers/{bin,lib,...}`.   So I think it is necessary to manually set `PATH`, etc.?

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: There is a brief description [here](https://docs.nvidia.com/hpc-sdk/hpc-sdk-install-guide/index.html).  

Essentially a network install allows the same install of the HPC SDK to be used on multiple systems with different properties.  On reflection, Spack already handles this, so I'm thinking the default should be changed to `single`?

**Comment by samcmill** [comment]
@Nischay-Pro I assume you are referring to Nsight Compute, since Nsight Systems does support Pascal.

Regardless, it would be incorrect to force a dependency on CUDA.  The HPC SDK provides a complete CUDA environment (it is not necessary to install the CUDA Toolkit to use the HPC SDK).  

In this extraordinary case, manually and separately installing the CUDA package is the best approach.

The CUDA variant flag should be used to mean that the CUDA bundled with the HPC SDK should be used instead of the `cuda` package.

**Comment by samcmill** [comment]
@scheibelp I'm going to need some help enabling a cuda variant.

I add the following the to nvhpc package:
```
variant('+cuda', default=True)
provides('cuda', when '+cuda')
```

But cuda from the HPC SDK is not getting picked up (note that's using the `cuda` package instead of `nvhpc`, but it does swap out `openblas` for `nvhpc` due to the `+blas`):
```
$ spack spec magma cuda_arch=70 ^nvhpc
Input spec
--------------------------------
magma cuda_arch=70
    ^nvhpc

Concretized
--------------------------------
magma@2.5.4%gcc@8.3.1+cuda+fortran+shared build_type=RelWithDebInfo cuda_arch=70 arch=linux-centos8-haswell
    ^cmake@3.18.2%gcc@8.3.1~doc+ncurses+openssl+ownlibs~qt arch=linux-centos8-haswell
        ^ncurses@6.2%gcc@8.3.1~symlinks+termlib arch=linux-centos8-haswell
            ^pkgconf@1.7.3%gcc@8.3.1 arch=linux-centos8-haswell
        ^openssl@1.1.1g%gcc@8.3.1+systemcerts arch=linux-centos8-haswell
            ^perl@5.30.3%gcc@8.3.1+cpanm+shared+threads arch=linux-centos8-haswell
                ^berkeley-db@18.1.40%gcc@8.3.1 arch=linux-centos8-haswell
                ^gdbm@1.18.1%gcc@8.3.1 arch=linux-centos8-haswell
                    ^readline@8.0%gcc@8.3.1 arch=linux-centos8-haswell
            ^zlib@1.2.11%gcc@8.3.1+optimize+pic+shared arch=linux-centos8-haswell
    ^cuda@11.1.0%gcc@8.3.1 arch=linux-centos8-haswell
        ^libxml2@2.9.10%gcc@8.3.1~python arch=linux-centos8-haswell
            ^libiconv@1.16%gcc@8.3.1 arch=linux-centos8-haswell
            ^xz@5.2.5%gcc@8.3.1~pic arch=linux-centos8-haswell
    ^nvhpc@20.9%gcc@8.3.1+blas+cuda+lapack+network~single arch=linux-centos8-haswell
```

If I add `provides('cuda')` to the `cuda` package, then I get `==> Error: Package '%s' cannot provide itself.`.

**Comment by samcmill** [comment]
@ajkotobi I've added a mpi variant.  

A quick test:
```
$ spack install netcdf-c%nvhpc ^nvhpc+mpi ^m4%gcc
```

I had issues with other dependent packages in the past, which is why I didn't originally include this in the first pass.  IIRC, the issue was that some dependent packages tried to use `spec['mpi'].prefix`, which can't simultaneously be the path to the MPI directory in the HPC SDK and the top level HPC SDK prefix.  Since the mpi variant needs more testing, I have set the default to False.

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/nvhpc/package.py:83]: Sorry I was assuming setting this environment variable changed where the installation placed the files. If I'm wrong there, then yes you would need to explicitly set `PATH` etc. in `setup_run_environment`.

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: Got it, in particular I see the following:

> This directory selection approach allows a network installation to support a network of machines running different versions of Linux. If all the platforms are identical, the shared installation location can perform a standard installation that all can use.

Based on this, I don't think the "network"-install type is ever needed: Spack will create a distinct prefix for every system that you install on (even if they all share the same networked filesystem installation prefix); for example Spack will create a distinct prefix for different distributions of Linux (Centos vs. Ubuntu) and also different major versions of the same distribution.

If the Nvidia HPC SDK installation requires finer granularity than (distro, major version), then I may have to think about some things that need adjusting in Spack to address that.

Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: Agreed.  I've kept the logic (just in case), but made single the default.

**Comment by scheibelp** [comment]
Regarding https://github.com/spack/spack/pull/19294#issuecomment-708470862, right now I think the path forward is not to specify `nvhpc` as a `cuda` provider:

* If users compiling with `nvhpc` should be using the `cuda` libs that it provides, then we currently don't support that seamlessly but I think we can advise users on how to make this work with a Spack environment.
* If users compiling with `nvhpc` can use any `cuda` libs, then it will work out fine not to add `depends_on(cuda)` or `provides(cuda)`: packages which specify a `cuda` dependency will pull in any implementation independently of the compiler used (including `nvphc`).
* If `nvhpc` provides additional libraries (besides `cuda`) and users would be expected to add `depends_on(nvhpc)` to their packages then there are other considerations, but I assume this is not the case. @samcmill @Nischay-Pro @ajkotobi please let me know if I'm wrong about that.

In more detail:

Spack doesn't allow a virtual package to exist with the same name as a real package, so you cannot add `provides(cuda)` to `nvhpc`. We would need to choose a new name like `cuda-virtual` (and packages would have to change their `depends_on` declarations to indicate that any provider of `cuda-virtual` is acceptable).

Whether this is currently worthwhile depends on how users are expected to make use of the `nvhpc` package. My original assumption was that users would install it exclusively to use it as a compiler, and that they generally wouldn't add it as a dependency. In that case, unfortunately Spack does not currently have a good mechanism to instruct dependents to use a specific `cuda` implementation when compiling with `nvhpc` (including the implementation provided by `nvhpc` itself).

I assume that packages which use the `Nvhpc` compiler and also depend on `cuda` should generally use the `cuda` libs provided by the installed `Nvhpc` package, but based on https://github.com/spack/spack/pull/19294#issuecomment-708459344 it sounds like it might "just work for now" if a separate `cuda` is installed? If there are version matching issues (e.g. a package using separate `cuda` libs needs a specific version) then I'll have to think of how to support that.

**Comment by samcmill** [comment]
All of the scenarios are possible.

- The `nvhpc` compilers can be used with an external CUDA install.  The forthcoming PR with package updates has some examples that do this.  There are version matching issues as `nvc++` is only recognized as a supported CUDA host compiler by external CUDA installs 11.0 and later.
- CUDA from the HPC SDK can be used with compilers other than the NV compilers (see the `nvhpc-byo-compiler` environment module included with the HPC SDK).
- The `nvhpc` compilers can of course be used with the bundled CUDA.

NVIDIA's strategy is for the NVIDIA HPC SDK is the "one stop shop" for HPC developers.  There is no need to combine it with a separately downloaded CUDA Toolkit.  In the future there may be capabilities in the HPC SDK that are not part of CUDA Toolkit.

Since exposing CUDA from the HPC SDK is a larger endeavor, I agree that a `cuda` variant should be deferred from the current PR.  However, the `cuda` dependency issue should be revisited so that `cuda` can also be provided by the `nvhpc` package.  (Aside: I wonder if renaming the `cuda` package to `cuda-toolkit`, for instance, and adding `provides('cuda')` to it would be a less painful change than `cuda-virtual`?)

Packages that indirectly use CUDA such as OpenACC or stdpar, can set `depends_on('nvhpc')` and the NV compilers will automatically include the appropriate bits just by setting the corresponding compiler flags (without needing to explicit expose `cuda` from `nvhpc` in Spack) and are satisfied by the current PR.

**Comment by samcmill** [comment]
> * If `nvhpc` provides additional libraries (besides `cuda`) and users would be expected to add `depends_on(nvhpc)` to their packages then there are other considerations, but I assume this is not the case. @samcmill @Nischay-Pro @ajkotobi please let me know if I'm wrong about that.

Besides CUDA, additional libraries include CPU BLAS and LAPACK libraries (already exposed as `blas` and `lapack` variants), as well as the NCCL and NVSHMEM communication libraries.

**Comment by scheibelp** [comment]
> (Aside: I wonder if renaming the cuda package to cuda-toolkit, for instance, and adding provides('cuda') to it would be a less painful change than cuda-virtual?)

Appropriating a name previously used for a package as a virtual has occurred before (e.g. for `pkg-config`) but generally it is good to avoid because preexisting installs can have trouble looking for package definitions once the package is renamed.

One thing I'm curious about: are these `blas`/`lapack`/`cuda` offerings fully independent implementations (e.g. written by the developers of `nvhpc`)? Or are they existing implementations bundled with the SDK? In the latter case, while it is not strictly wrong to model `nvhpc` as a `cuda...` provider it may be confusing to users interfacing the Spack repository to see it modeled this way.

> the cuda dependency issue should be revisited so that cuda can also be provided by the nvhpc package. 

To make sure: you want to resolve this before this PR is merged? Or are you saying it's worth resolving "soon"?

Another question: the `Nvhpc.install` method invokes the `install` script: does this do compilation from source etc. or is it a relocation of binaries in the downloaded tarball?

**Comment by samcmill** [comment]
> One thing I'm curious about: are these `blas`/`lapack`/`cuda` offerings fully independent implementations (e.g. written by the developers of `nvhpc`)? Or are they existing implementations bundled with the SDK? In the latter case, while it is not strictly wrong to model `nvhpc` as a `cuda...` provider it may be confusing to users interfacing the Spack repository to see it modeled this way.

The latter.
 
> > the cuda dependency issue should be revisited so that cuda can also be provided by the nvhpc package.
> 
> To make sure: you want to resolve this before this PR is merged? Or are you saying it's worth resolving "soon"?

Sorry for not being clearer.  I'm suggesting leaving the cuda issue unresolved in this PR and to merge without cuda.  There seem to some fundamental questions that need to be answered first and it does not make sense to me to hold up the rest while that's being investigated.

> Another question: the `Nvhpc.install` method invokes the `install` script: does this do compilation from source etc. or is it a relocation of binaries in the downloaded tarball?

The latter, relocation of binaries and other files.

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: It occurred to me that this could be simplified in one of two ways:

Firstly, since there are two boolean variants which must have opposing values, you could just have one boolean variant (e.g. keep `network` and set it to `False` by default) - in that case the `if/elif` in `install` could be an `if/else`.

If you think there is value in explicitly labelling each state, you can also encode this as a single valued variant. The `dims` variant in `boxlib` is an example of this:

```
    variant('dims', default='3', values=('1', '2', '3'), multi=False,
            description='Number of spatial dimensions')
```

In this case it would look like

```
    variant('install-type', default='single', values=('single, 'network'), multi=False,
            description='Network installs are for installations shared by different operating systems')
```

**Review by scheibelp — APPROVED** [review]
I have one suggestion for a simplification but with this patch I can install and use the `nvhpc` compiler.

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (samcmill) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: Great idea.  I changed it to use the single valued variant approached.

File: lib/spack/spack/compilers/nvhpc.py
Comment (alalazo) [inline lib/spack/spack/compilers/nvhpc.py:48]: If we don't need to compute the string based on the object state, can't we just do:
```suggestion
    openmp_flag = "-mp"
```
for all the properties?

**Review by adamjstewart — APPROVED** [review]
LGTM, but I would like to see approvals from @RemiLacroix-IDRIS @ajkotobi @pramodk @fspiga @Nischay-Pro if possible.

File: lib/spack/spack/compilers/nvhpc.py
Comment (samcmill) [reply lib/spack/spack/compilers/nvhpc.py:48]: It could.  Currently it's just following the convention used by the most of the compilers.  

For `openmp_flag` in particular:
- `apple_clang` and `clang` directly set `openmp_flag`
- `cce` and `intel` depend on the object state
- the rest use the `@property` decorated getter without depending on the object state.

Comment (adamjstewart) [reply lib/spack/spack/compilers/nvhpc.py:48]: I'm fine with either syntax. Technically they aren't identical, as a getter without a setter won't let you set, but I don't think we've ever needed that functionality.

**Comment by Nischay-Pro** [comment]
PR looks good to me.

**Review by Nischay-Pro — APPROVED** [review]

File: lib/spack/spack/compilers/nvhpc.py
Comment (samcmill) [reply lib/spack/spack/compilers/nvhpc.py:48]: > Technically they aren't identical, as a getter without a setter won't let you set

I lean toward keeping it as is for this reason, but I'm happy to change it if @alalazo feels strongly about it.

**Comment by ajkotobi** [comment]
Looks good to me

**Review by ajkotobi — APPROVED** [review]

**Comment by scheibelp** [comment]
Fantastic - thanks for looking at this @Nischay-Pro @ajkotobi @adamjstewart! I think if there are concerns at this point we can address them in a follow-up PR.

**Pull request merged**
